hdimage: Add GPT support

ssorensen opened the pull request:
Fixes #40.

tpetazzoni commented:
It would be very useful to have this feature merged in genimage.

michaelolbrich commented on crc32.c:
```diff
@@ -0,0 +1,60 @@
+		crc = crc32_tab[(crc ^ *p++) & 0xff] ^ (crc >> 8);
+
+	return ~crc;
+}
```
Hmmm, I don't really like this, but adding some big dependency just for crc32 isn't nice either.

michaelolbrich commented on image-hd.c:
```diff
@@ -166,6 +199,160 @@ static int hdimage_insert_ebr(struct image *image, struct partition *part)
+	if (hd->disk_uuid) {
+		if (uuid_parse(hd->disk_uuid, header.disk_uuid) == -1) {
+			image_error(image, "invalid disk UUID: %s\n", hd->disk_uuid);
+			return -EINVAL;
```
This should be in setup

michaelolbrich commented on image-hd.c:
```diff
@@ -166,6 +199,160 @@ static int hdimage_insert_ebr(struct image *image, struct partition *part)
+		}
+		else {
+			uuid_random(table[i].uuid);
+		}
```
Also in setup. Any parsing / validating of the config should be handled there.
Basically only I/O errors should cause generate to fail,

michaelolbrich commented on image-hd.c:
```diff
@@ -166,6 +199,160 @@ static int hdimage_insert_ebr(struct image *image, struct partition *part)
+	if (ret) {
+		image_error(image,
+			    "failed to write image partition '%s'\n",
+			    part->name);
```
This error message makes no sense. There is no partition here. We're expanding the image to the full size here, right?

michaelolbrich commented on image-hd.c:
```diff
@@ -166,6 +199,160 @@ static int hdimage_insert_ebr(struct image *image, struct partition *part)
+	ret = insert_data(image, (char *)&table, outfile, sizeof(table),
+			  image->size - GPT_SECTORS*512);
+	if (ret) {
+		image_error(image, "failed to write GPT table\n");
```
This should be `... write backup GPT table`, right?

michaelolbrich commented on test/basic-images.test:
```diff
@@ -214,6 +214,18 @@ test_expect_success "hdimage2" "
 	test_must_fail run_genimage hdimage2.config test.hdimage
 "
 
+test_expect_success fdisk,sfdisk "hdimage4" "
```
This fails with the old test setup, right? We should skip the test for old versions of `fdisk` / `sfdisk`.
Untested:
```
sfdisk -h | grep -q gpt && test_set_prereq sfdisk-gpt
fdisk -h | grep -q gpt && test_set_prereq fdisk-gpt
test_expect_success fdisk-gpt,sfdisk-gpt "hdimage4" "
...
```

michaelolbrich commented on image-hd.c:
```diff
@@ -166,6 +199,160 @@ static int hdimage_insert_ebr(struct image *image, struct partition *part)
 }
 
+static const char *
+gpt_partition_type_lookup(const char *str)
```
The string-length check is outside of this function, so I think the argument should just be the single `char`.

ssorensen commented:
I have updated the code:
  * Moved all validation to setup
  * Fixed copy/paste errors in messages
  * Updated the test to check for GPT capable fdisk
  * Added seeding of the RNG so the random uuids are actually rando

lp35 commented:
Wonderful work! Thank you very much !!! 👍
